payments-core/src/main/java/com/stripe/android/analytics/SessionSavedStateHandler.kt

```diff
@@ -0,0 +1,77 @@
+package com.stripe.android.analytics
+
+import androidx.annotation.RestrictTo
+import androidx.lifecycle.SavedStateHandle
+import com.stripe.android.core.networking.AnalyticsRequestFactory
+import org.jetbrains.annotations.VisibleForTesting
+import java.util.UUID
+
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+object SessionSavedStateHandler {
+    @VisibleForTesting
+    internal const val SESSION_KEY = "STRIPE_ANALYTICS_LOCAL_SESSION"
+
+    private var sessionLocked: Boolean = false
+
+    fun startSession(savedStateHandle: SavedStateHandle) {
+        savedStateHandle.get<Session>(SESSION_KEY)?.let { session ->
+            session.also { storedSession ->
+                when (storedSession) {
+                    is Session.Owner -> {
+                        AnalyticsRequestFactory.setSessionId(UUID.fromString(storedSession.id))
+                        sessionLocked = true
+                    }
+                    is Session.Observer -> Unit
+                }
+            }
+        } ?: run {
+            val session = if (!sessionLocked) {
+                sessionLocked = true
+
+                val id = UUID.randomUUID()
+
+                AnalyticsRequestFactory.setSessionId(id)
+
+                Session.Owner(id.toString())
+            } else {
+                Session.Observer
+            }
+
+            savedStateHandle[SESSION_KEY] = session
+        }
+    }
+
+    fun restartSession(savedStateHandle: SavedStateHandle) {
+        savedStateHandle.get<Session>(SESSION_KEY)?.let { session ->
+            session.also { storedSession ->
+                when (storedSession) {
+                    is Session.Owner -> {
+                        val id = UUID.randomUUID()
+
+                        AnalyticsRequestFactory.setSessionId(id)
+
+                        savedStateHandle[SESSION_KEY] = Session.Owner(id.toString())
+                    }
+                    is Session.Observer -> Unit
+                }
+            }
+        }
+    }
+
+    fun clearSession(savedStateHandle: SavedStateHandle) {
+        savedStateHandle.get<Session>(SESSION_KEY)?.let { session ->
```

Comment:
Do we need to call `remove()` here instead of `get()`?

Reply:
Ideally yes, but since we have no way of telling if a `ViewModel` is being closed from process death or if the user manually exited the lifecycle component, we have to simply check if we can unlock session. Otherwise, calling `remove` will also remove the stored data in process death cases.